### PR TITLE
Add lifetime encoding at end of block

### PR DESCRIPTION
**Pull request by pascal-huber**
# Work

 - [x] Count number of lifetimes for read permission amount
 - [x] Use NewLft statements for lifetimes needed by the target block of the goto-terminator
 - [x] Implement and use ShortenLifetime statement to shorten derived lifetimes as needed by the target block of the goto-terminator

# Example

Here the if-branch creates lifetimes `lft6 ⊆ bw0`, the else-branch creates `lft6 ⊆ bw1`. After the branch we require the lifetimes to be the same. This means that the if-branch also needs to create `bw1` and the else-branch also needs to create `bw0` and `lft6` must be shortened to  `lft6 ⊆ bw0 ∩ bw1`.

```
fn main(c: bool){
    let mut a: i32 = 4;
    let mut b: i32 = 5;
    let x;
    if c {
        x = &mut a;
    } else {
        x = &mut b;
    }
    *x = 6;
}
```

### CODE REVIEW

File: prusti-viper/src/encoder/mir/procedures/encoder.rs [hunk 1]
```diff
@@ -104,7 +104,8 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
             deallocate_returns,
             assert_postconditions,
         );
-        self.encode_body(&mut procedure_builder)?;
+        let rd_perm: u32 = self.lifetimes.lifetime_count();
```
Comment (vakaras) [inline prusti-viper/src/encoder/mir/procedures/encoder.rs]: Why is this not a field on `ProcedureEncoder`, but an additional argument/parameter that needs to be threaded through?

File: vir/defs/high/ast/statement.rs [hunk 2]
```diff
@@ -188,3 +189,11 @@ pub struct GhostAssignment {
     pub value: Expression,
     pub position: Position,
 }
+
+#[display(fmt = "{} := shorten_lifetime({:?}, {})", target, value, rd_perm)]
+pub struct ShortenLifetime {
+    pub target: VariableDecl,
+    pub value: Expression,
```
Comment (vakaras) [inline vir/defs/high/ast/statement.rs]: Is this statement supposed to intersect a series of lifetimes and while doing so modify their `LifetimeTokens`? If yes, then using `Vec<LifetimeConst>` would make it easier to generate `vir_low` than `Expression`.

**Review by vakaras — CHANGES_REQUESTED** [review]
LGTM, except for the two comments I added.

File: vir/defs/high/operations_internal/identifier/ty.rs [hunk 3]
```diff
@@ -128,7 +128,11 @@ impl WithIdentifier for ty::Slice {
 
 impl WithIdentifier for ty::Reference {
     fn get_identifier(&self) -> String {
-        format!("ref${}", self.target_type.get_identifier())
+        format!(
+            "ref$&{}${}",
+            self.lifetime.get_identifier(),
+            self.target_type.get_identifier(),
+        )
```
Comment (pascal-huber) [inline vir/defs/high/operations_internal/identifier/ty.rs]: @vakaras this change may be causing some problems, looking into it

File: prusti-viper/src/encoder/mir/procedures/encoder.rs (hunk 1 above)
Comment (pascal-huber) [reply prusti-viper/src/encoder/mir/procedures/encoder.rs]: yes, done.

File: vir/defs/high/ast/statement.rs (hunk 2 above)
Comment (pascal-huber) [reply vir/defs/high/ast/statement.rs]: I think you're right, this makes no sense. In vir_low I want to create something like this:

```
var old_lft6: Lifetime := lft6
lft6 := lft_tok_sep_take(bw0, bw1, q)
bor_shorten(lft6, old_lft6, q, _4.ref)
```

Having a vector would be better here. And `_4.ref` we will have to look up in vir_low, right?

Comment (vakaras) [reply vir/defs/high/ast/statement.rs]: Does `ShortenLifetime` correspond to `lft_tok_sep_take`, to `bor_shorten`, or to both? If `bor_shorten` is needed, then I would say that you have to include `_4` into the struct.

Comment (pascal-huber) [reply vir/defs/high/ast/statement.rs]: Originally I was intending to do `lft_tok_sep_take` only - but doing both at the same time seems nicer. I think some work will be required to include `_4`, not sure.

Comment (pascal-huber) [reply vir/defs/high/ast/statement.rs]: renamed to `LifetimeTake`.

**Review by vakaras — APPROVED** [review]
LGTM

bors r+

**Pull request merged**